src/Stratis.Bitcoin.Features.Consensus.Tests/TestChainFactory.cs

```diff
@@ -120,34 +125,70 @@ internal class TestChainFactory
             List<Block> blocks = new List<Block>();
             for (int i = 0; i < count; ++i)
             {
-                PowBlockAssembler blockAssembler = CreatePowBlockAssembler(testChainContext.Network, testChainContext.Consensus, testChainContext.Chain, mempoolLock, mempool, testChainContext.DateTimeProvider, testChainContext.LoggerFactory as LoggerFactory);
+                BlockTemplate newBlock = await MineBlockAsync(testChainContext,
+                    scriptPubKey, mempoolLock, mempool, mutateLastBlock && i == count - 1);
+                blocks.Add(newBlock.Block);
+            }
 
-                BlockTemplate newBlock = blockAssembler.CreateNewBlock(scriptPubKey);
+            return blocks;
+        }
 
-                int nHeight = testChainContext.Chain.Tip.Height + 1; // Height first in coinbase required for block.version=2
-                Transaction txCoinbase = newBlock.Block.Transactions[0];
-                txCoinbase.Inputs[0] = TxIn.CreateCoinbase(nHeight);
-                newBlock.Block.UpdateMerkleRoot();
+        private static async Task<BlockTemplate> MineBlockAsync(TestChainContext testChainContext, Script scriptPubKey,
+            MempoolSchedulerLock mempoolLock, TxMempool mempool, bool getMutatedBlock = false)
+        {
+            BlockTemplate newBlock = CreateBlockTemplate(testChainContext, scriptPubKey, mempoolLock, mempool);
 
-                var maxTries = int.MaxValue;
+            if(getMutatedBlock) newBlock.Block.Transactions.Add(newBlock.Block.Transactions[0]);
```

Comment:
You are duplicating a coinbase tx. 

There can be only one coinbase and coinstake tx so the block fails several consensus rules, not just mutated merkle tree one.

In order to create a mutated block that exploits CVE-2012-2459 you need to create a fully valid block with a few normal tx and duplicate 2 of them, after that update merkle root and check if it's the same as before the duplication.

Reply:
That is probably true, I think I will try to do exactly that, but for the purpose of these tests I think they are already good enough to prove the code works as expected 
![image](https://user-images.githubusercontent.com/4638821/39420179-53a04436-4c5b-11e8-8504-9c3bc17f4b64.png)
The tx-duplicate is the first error appearing in the block validation and the code goes to the correct path during the test. I will try to build a more realistic block today, but if we don't decide that rules can be run in any order, I think this is already a proof no?

Reply:
You may be right but rules are such an important part of the code I would suggest to try to create proper invalid merkle trx.